File: src/main/java/com/github/jactor/web/interceptor/RequestInterceptor.java
package com.github.jactor.web.interceptor;

import com.github.jactor.web.Language;
import com.github.jactor.web.Request;
import com.github.jactor.web.RequestManager;
import java.util.List;
import java.util.Locale;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.PropertySource;
import org.springframework.context.i18n.LocaleContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.ModelAndView;

@Component
@PropertySource("classpath:application.properties")
public class RequestInterceptor implements HandlerInterceptor {

  static final String CHOSEN_LANGUAGE = "chosenLanguage";
  static final String CURRENT_REQUEST = "currentRequest";

  private static final Language LANGUAGE_DEFAULT_IS_ENGLISH = new Language(new Locale("en"), "English");
  private static final List<Language> SUPPORTED_LANGUAGES = List.of(
      LANGUAGE_DEFAULT_IS_ENGLISH,
      new Language(new Locale("no"), "Norsk"),
      new Language(new Locale("th"), "ไทย")
  );

  private final String contextPath;

  @Autowired
  public RequestInterceptor(@Value("${server.servlet.context-path}") String contextPath) {
    this.contextPath = contextPath;
  }

  @Override
  public void postHandle(
      HttpServletRequest servletRequest,
      HttpServletResponse servletResponse,
      Object handler,
      ModelAndView modelAndView
  ) {
    RequestManager requestManager = new RequestManager(contextPath, servletRequest);
    Request request = new Request(requestManager.fetchCurrentUrl(), requestManager.fetchChosenView());
    Language chosenLanguage = fetchChosenLangugae(requestManager);

    modelAndView.getModelMap().addAttribute(CHOSEN_LANGUAGE, chosenLanguage);
    modelAndView.getModelMap().addAttribute(CURRENT_REQUEST, request);
  }

  private Language fetchChosenLangugae(RequestManager requestManager) {
    if (requestManager.noLanguageParameters()) {
      return requestManager.fetchFrom(SUPPORTED_LANGUAGES, LocaleContextHolder.getLocale(), LANGUAGE_DEFAULT_IS_ENGLISH);
    }

    return requestManager.fetchFromParameters(SUPPORTED_LANGUAGES, LANGUAGE_DEFAULT_IS_ENGLISH);
  }
}

File: web/src/main/kotlin/com/github/jactor/web/interceptor/RequestInterceptor.kt
package com.github.jactor.web.interceptor

import com.github.jactor.web.Language
import com.github.jactor.web.Request
import com.github.jactor.web.RequestManager
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.PropertySource
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor
import org.springframework.web.servlet.ModelAndView
import java.util.Locale

@Component
@PropertySource("classpath:application.properties")
class RequestInterceptor(@param:Value("\${server.servlet.context-path}") private val contextPath: String) : HandlerInterceptor {

    companion object {
        const val CHOSEN_LANGUAGE = "chosenLanguage"
        const val CURRENT_REQUEST = "currentRequest"

        @JvmStatic
        private val LANGUAGE_DEFAULT_IS_ENGLISH = Language(Locale("en"), "English")

        @JvmStatic
        private val SUPPORTED_LANGUAGES = listOf(
            LANGUAGE_DEFAULT_IS_ENGLISH,
            Language(Locale("no"), "Norsk"),
            Language(Locale("th"), "ไทย")
        )
    }

    override fun postHandle(
        servletRequest: HttpServletRequest,
        servletResponse: HttpServletResponse,
        handler: Any,
        modelAndView: ModelAndView?
    ) {
        if (modelAndView != null) {
            val requestManager = RequestManager(contextPath, servletRequest)
            val request = Request(requestManager.fetchCurrentUrl(), requestManager.fetchChosenView())
            val chosenLanguage = fetchChosenLangugae(requestManager)

            modelAndView.modelMap.addAttribute(CHOSEN_LANGUAGE, chosenLanguage)
            modelAndView.modelMap.addAttribute(CURRENT_REQUEST, request)
        }
    }

    private fun fetchChosenLangugae(requestManager: RequestManager): Language {
        return if (requestManager.noLanguageParameters()) {
            requestManager.fetchFrom(SUPPORTED_LANGUAGES, LocaleContextHolder.getLocale(), LANGUAGE_DEFAULT_IS_ENGLISH)
        } else requestManager.fetchFromParameters(SUPPORTED_LANGUAGES, LANGUAGE_DEFAULT_IS_ENGLISH)
    }
}